File: src/test/java/com/kata/bank_account/domain/entities/SavingAccountTest.java
package com.kata.bank_account.domain.entities;

import com.kata.bank_account.domain.exception.AboveCeilingException;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class SavingAccountTest {

    @Test
    void should_have_ceiling() {

        BigDecimal ceiling = BigDecimal.valueOf(100);
        var savingAccount = new SavingAccount(ceiling);

        assertThat(savingAccount.ceiling()).isEqualTo(ceiling);
    }

    @Test
    void should_throw_on_deposit_greater_than_ceiling() {
        BigDecimal ceiling = BigDecimal.valueOf(100);
        var savingAccount = new SavingAccount(ceiling);
        BigDecimal depositAmount = ceiling.multiply(BigDecimal.TEN);

        assertThatThrownBy(() -> savingAccount.deposit(depositAmount))
                .isInstanceOf(AboveCeilingException.class);
    }

    @Test
    void should_allow_deposit_within_ceiling() {
        BigDecimal ceiling = BigDecimal.valueOf(100);
        var savingAccount = new SavingAccount(ceiling);

        BigDecimal depositAmount = BigDecimal.TEN;
        savingAccount.deposit(depositAmount);

        assertThat(savingAccount.balance()).isEqualTo(depositAmount);

    }

    @Test
    void should_not_have_overdraft() {
        BigDecimal ceiling = BigDecimal.valueOf(100);
        var savingAccount = new SavingAccount(ceiling);

        assertThat(savingAccount.canOverdraft()).isFalse();

    }
}

File: src/test/kotlin/com/kata/bank/domain/model/SavingAccountTest.kt
package com.kata.bank.domain.model


import com.kata.bank.domain.exception.AboveCeilingException
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test

class SavingAccountTest {

    @Test
    fun should_not_have_overdraft() {
        val ceiling: Long = 100
        val savingAccount = SavingAccount(ceiling)

        assertThat(savingAccount.overDraft).isEqualTo(0)
    }

    @Test
    fun should_have_ceiling() {
        val ceiling: Long = 100
        val savingAccount = SavingAccount(ceiling)

        assertThatThrownBy { savingAccount.deposit(200) }
            .isInstanceOf(AboveCeilingException::class.java)
    }

    @Test
    fun should_allow_deposit_within_ceiling() {
        val ceiling: Long = 100
        val savingAccount = SavingAccount(ceiling)

        savingAccount.deposit(50)

        assertThat(savingAccount.balance()).isEqualTo(50)
    }
}